### PR TITLE
Add authentication and loading gists

### PR DESCRIPTION
**Pull request by algomaster99**
Here is how the authentication button will look:
auth

The PR will be ready for merge after:
- [x] The profile image is sent by back-end
- [ ] I have currently removed the docs tab. Needs to be adjusted somewhere.

### CODE REVIEW

**Comment by algomaster99** [comment]
@domoritz Please fetch this and test. The gif didn't upload correctly :confused:

**Comment by domoritz** [comment]
Instead of using the three dots, can we just make the profile image the button? Can you also make the popup a bit larger so it looks nicer/ more consistent?

File: src/components/header/renderer.tsx [hunk 1]
```diff
@@ -19,10 +20,13 @@ type Props = ReturnType<typeof mapStateToProps> &
   ReturnType<typeof mapDispatchToProps> & { history: any; showExample: boolean };
 
 interface State {
+  open: boolean;
   showVega: boolean;
   scrollPosition: number;
 }
 
+const url = 'https://vega.now.sh/';
```
Comment (domoritz) [inline src/components/header/renderer.tsx]: Use all caps and use a more descriptive name.

File: src/components/header/renderer.tsx [hunk 2]
```diff
@@ -171,6 +213,25 @@ class Header extends React.PureComponent<Props, State> {
       </div>
     );
 
+    const auth = this.props.isAuthenticated ? (
```
Comment (domoritz) [inline src/components/header/renderer.tsx]: Can we make the profile button a new component?

File: src/components/header/renderer.tsx [hunk 3]
```diff
@@ -434,11 +535,9 @@ class Header extends React.PureComponent<Props, State> {
             ]}
           </PortalWithState>
 
-          <span>{docsLink}</span>
+          {/* <span>{docsLink}</span> */}
```
Comment (domoritz) [inline src/components/header/renderer.tsx]: Can we leave this in for now and reduce the spacing between buttons?

**Comment by algomaster99** [comment]
@domoritz 
### Logged In
Screenshot from 2019-07-12 21-32-18

### Logged Out
Screenshot from 2019-07-12 22-24-47

**Comment by algomaster99** [comment]
@domoritz Do you have any idea why travis is failing?

**Comment by domoritz** [comment]
Something is broken. I'm looking.

**Comment by domoritz** [comment]
The "log out" menu doesn't look like the drop down for the other menu items. The log in with black background looks a bit out of place. Have a look at https://observablehq.com/ for inspiration.

File: src/components/header/renderer.tsx [hunk 4]
```diff
@@ -165,6 +203,35 @@ class Header extends React.PureComponent<Props, State> {
...
+                    <input className="sign-out" type="submit" value="Logout" onClick={e => e.stopPropagation()} />
+                  </div>
+                </div>
+              )}
+            </div>
+          </form>
+        ) : (
+          <form action={`${BACKEND_URL}auth/github`} method="get">
+            <button className="sign-in" type="submit">
+              <span className="sign-in-text">Login with</span>
```
Comment (domoritz) [inline src/components/header/renderer.tsx]:
```suggestion
              <span className="sign-in-text">Sign in with</span>
```

File: src/components/header/renderer.tsx [hunk 5]
```diff
@@ -165,6 +203,35 @@ class Header extends React.PureComponent<Props, State> {
       </div>
     );
 
+    const authButton = (
+      <div className="auth-button-container">
+        {this.props.isAuthenticated ? (
+          <form action={`${BACKEND_URL}auth/github/logout`} method="get">
+            <div className="profile-container">
```
Comment (domoritz) [inline src/components/header/renderer.tsx]: That's a lot of classes. Do we need them all?

File: src/constants/consts.ts [hunk 6]
```diff
@@ -53,3 +53,7 @@ export const KEYCODES = {
   SINGLE_QUOTE: 222,
   SLASH: 191,
 };
+
+export const BACKEND_URL = 'https://vega.now.sh/';
+
+export const COOKIE_NAME = 'vegasessid';
```
Comment (domoritz) [inline src/constants/consts.ts]:
```suggestion
export const COOKIE_NAME = 'vega_session';
```

File: src/constants/default-state.ts [hunk 7]
```diff
@@ -5,6 +5,7 @@ import { LocalLogger } from '../utils/logger';
 import { LAYOUT, Mode, NAVBAR, Renderer, SIDEPANE, VEGA_LITE_START_SPEC, View } from './consts';
 
 export interface State {
+  isAuthenticated: boolean;
```
Comment (domoritz) [inline src/constants/default-state.ts]: Can we even put this in the state that is being serialized? I feel we might need to check this when the user is opening the editor. Or am I missing something?

File: src/utils/getCookie.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+const getCookie = cname => {
```
Comment (domoritz) [inline src/utils/getCookie.ts]: Let's make this a real function.

File: src/components/header/index.css [hunk 9]
```diff
@@ -297,8 +297,123 @@
...
+.welcome {
+  font-size: 0.7em;
+  color: #6e6e6e;
+  margin: 1em 0 0.5em 0.6em;
+}
+
+.whoami {
+  font-size: 0.9em;
+  padding: 0.5em 0 1em 0.5em;
+  /* padding-bottom: 0.5em; */
```
Comment (domoritz) [inline src/components/header/index.css]: delete

**Comment by domoritz** [comment]
- [ ] Add description to the readme about how to test this feature.

**Comment by domoritz** [comment]
- [ ] Make sure that we don't have unsaved changes before navigating away. Can we somehow sign in in a new tab rather than reloading the page?

File: src/components/header/renderer.tsx [hunk 10]
```diff
@@ -165,6 +203,35 @@ class Header extends React.PureComponent<Props, State> {
...
+              <img className="profile-img" src={this.props.profilePicUrl} />
+              <span className="arrow-down"></span>
+              {this.state.open && (
+                <div className="profile-menu">
+                  <div className="welcome">Welcome!</div>
+                  <div className="whoami">{this.props.name}</div>
+                  <div className="profile-options">
+                    <input className="sign-out" type="submit" value="Logout" onClick={e => e.stopPropagation()} />
```
Comment (domoritz) [inline src/components/header/renderer.tsx]:
```suggestion
                    <input className="sign-out" type="submit" value="Sign out" onClick={e => e.stopPropagation()} />
```

File: src/components/header/renderer.tsx [hunk 11]
```diff
@@ -165,6 +203,35 @@ class Header extends React.PureComponent<Props, State> {
...
+    const authButton = (
+      <div className="auth-button-container">
+        {this.props.isAuthenticated ? (
+          <form action={`${BACKEND_URL}auth/github/logout`} method="get">
+            <div className="profile-container">
+              <img className="profile-img" src={this.props.profilePicUrl} />
+              <span className="arrow-down"></span>
+              {this.state.open && (
+                <div className="profile-menu">
+                  <div className="welcome">Welcome!</div>
```
Comment (domoritz) [inline src/components/header/renderer.tsx]:
```suggestion
                  <div className="welcome">Logged in as</div>
```

**Comment by domoritz** [comment]
Screen Shot 2019-08-11 at 1 55 16 PM

* Make the rule less prominent
* Add more padding around the popup
* Move the popup up
* Make it somehow clear (visually) that you cannot click on the user name

**Comment by domoritz** [comment]
Clicking on the user name should open the model that shows the user's visualizations (the gist dialog for now).

File: src/constants/default-state.ts (hunk 7 above)
Comment (algomaster99) [reply src/constants/default-state.ts]: @domoritz Yes, we check this. Doesn't it automatically store it in localStorage?

Comment (domoritz) [reply src/constants/default-state.ts]: It does but my question is whether that is right. If we serialize this and then the user opens the editor after 10 years, we will tell them that they are logged in even if the login has timed out.

File: src/components/header/renderer.tsx (hunk 5 above)
Comment (algomaster99) [reply src/components/header/renderer.tsx]: @domoritz yeah, the component is a bit complicated

File: src/constants/default-state.ts (hunk 7 above)
Comment (algomaster99) [reply src/constants/default-state.ts]: @domoritz Whenever the editor is opened, a fetch request is made to the back-end to check if the user is logged in and accordingly the state is set.

File: src/utils/getCookie.ts (hunk 8 above)
Comment (algomaster99) [reply src/utils/getCookie.ts]: @domoritz Can you elaborate? I don't get you. :/

Comment (domoritz) [reply src/utils/getCookie.ts]: If you can, use `function` instead of making this a `const gteCookie`.

File: src/constants/default-state.ts (hunk 7 above)
Comment (domoritz) [reply src/constants/default-state.ts]: Okay, then let's not persists `isAuthenticated`. Can you exclude it?

**Comment by algomaster99** [comment]
@domoritz We can sign-in in a new tab but what are the unsaved changes? Editor most of the times save things automatically.

**Comment by domoritz** [comment]
For example, the interactive state of a visualization is not saved.

File: src/components/header/renderer.tsx (hunk 5 above)
Comment (algomaster99) [reply src/components/header/renderer.tsx]: @domoritz I removed a class called `profile-options`.

**Comment by domoritz** [comment]
- [ ] Don't make the logout text blue

**Comment by domoritz** [comment]
Missing fix for

Screen Shot 2019-08-18 at 10 16 30 AM

**Pull request merged**